### ISSUE TITLE
custom stopping_critriea function doesn't receive logits scores (receives None instead)

### ISSUE THREAD

**Issue by Gandalf098** [huggingface/transformers#23674]
### System Info

- `transformers` version: 4.29.2
- Platform: Linux-5.15.107+-x86_64-with-glibc2.31
- Python version: 3.10.11
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Tensorflow version (GPU?): 2.12.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.6.9 (gpu)
- Jax version: 0.4.8
- JaxLib version: 0.4.7
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Reproduction Steps:

1. Initialize a BART model & its tokenizer (in my case it is facebook/bart-large)
2. Create a custom stopping_criteria function and add it to StoppingCriteriaList object
3. Run model.generate() with the your stopping criteria list as argument

Scores argument is always None

Example code:
```python
import torch
from transformers import StoppingCriteriaList, BartForConditionalGeneration, BartTokenizer

def custom_stopping_criteria(input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
    print ("Scores:", scores)
    return False

stopping_criteria = StoppingCriteriaList([custom_stopping_criteria])

model = BartForConditionalGeneration.from_pretrained("facebook/bart-large", forced_bos_token_id=0)
tok = BartTokenizer.from_pretrained("facebook/bart-large")

example_english_phrase = "UN Chief Says There Is No <mask> in Syria"
batch = tok(example_english_phrase, return_tensors="pt")

model.generate(batch["input_ids"], stopping_criteria=stopping_criteria)
```

The above code uses a stopping critriea that just prints the scores value when called (which prints None)

### Expected behavior

The expected behavior should be to have Scores logits populated with values instead of being None (values before or after softmax don't matter)

**Comment by sgugger** [comment]
cc @gante

**Comment by gante** [comment]
Hey @Gandalf098 (the white, I hope ;) )

By default, the scores are not initialized and are kept as `None` (see [here](https://github.com/huggingface/transformers/blob/5fa0a1b23b6a79eb646636dbf9a22cb34ff48a74/src/transformers/generation/utils.py#L2308)). To enable score-keeping, you must pass `return_dict_in_generate=True, output_scores=True` to your `.generate()` call.

____________________________________________

```py
import torch
from transformers import StoppingCriteriaList, BartForConditionalGeneration, BartTokenizer

def custom_stopping_criteria(input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
    print("Scores:", scores)
    return False

stopping_criteria = StoppingCriteriaList([custom_stopping_criteria])

model = BartForConditionalGeneration.from_pretrained("facebook/bart-large", forced_bos_token_id=0)
tok = BartTokenizer.from_pretrained("facebook/bart-large")

example_english_phrase = "UN Chief Says There Is No <mask> in Syria"
batch = tok(example_english_phrase, return_tensors="pt")

model.generate(batch["input_ids"], stopping_criteria=stopping_criteria, return_dict_in_generate=True, output_scores=True)
```

**Comment by ylacombe** [comment]
Hi @gante and @Gandalf098,

According to the `StoppingCriteria.__call__` [signature](https://huggingface.co/docs/transformers/v4.30.0/en/internal/generation_utils#transformers.StoppingCriteria) and to its docstring, `scores` is supposed to be a `torch.FloatTensor`.
> scores (torch.FloatTensor of shape (batch_size, config.vocab_size)) — Prediction scores of a language modeling head.


It makes sense to think of it as the **last** prediction scores of the language modeling head, meaning that the score-keeping here refers not to `score` (optional history of the prediction scores) but to `next_token_scores` (always available last prediction scores - at least for [greedy decoding](https://github.com/huggingface/transformers/blob/v4.30.0/src/transformers/generation/utils.py#L2400-L2401), we should verify for other decoding strategies). 

In that sense, I do think we should correct this point. What do you think @gante?

**Comment by qgallouedec** [comment]
Hey @Gandalf098, apologies for the very late follow-up, we're catching up on a long backlog.

This is working as intended: `scores` is only tracked when you ask for it, because keeping the full score history costs real memory (~50MB per 1000 tokens for a 7B model in bf16), so passing `return_dict_in_generate=True, output_scores=True` to `generate()` populates it, exactly as @gante showed above.
The docs gap that kept this issue open was closed by #26863: the requirement is now spelled out on `StoppingCriteria` and on `generate`'s `stopping_criteria` argument https://github.com/huggingface/transformers/blob/fdbd8b06fc39ecf5fbf94b62c1dcbb1b871ddc6c/src/transformers/generation/stopping_criteria.py#L31-L34

Closing as intended behavior; this also duplicates #18068. One genuine leftover we'll track separately: the docstring still types `scores` as a single `torch.FloatTensor` of shape `(batch_size, config.vocab_size)`, whereas what is actually passed is the growing tuple of per-step score tensors, the point @ylacombe raised above.
